target_chains/ton/contracts/contracts/Pyth.fc

```diff
@@ -198,34 +211,54 @@ tuple parse_price_feeds_from_data(int msg_value, slice data, tuple price_ids, in
 
     int root_digest = parse_pyth_payload_in_wormhole_vm(payload);
 
-    ;; Create tuple to store price feeds
-    tuple price_feeds = empty_tuple();
+    ;; Create dictionary to store price feeds in order (dict has a udict_get_next? method which returns the next key in order)
+    cell ordered_feeds = new_dict();
     ;; Track which price IDs we've found
     cell found_price_ids = new_dict();
 
+    int index = 0;
+
     repeat(num_updates) {
         (int price_id, int price, int conf, int expo, int publish_time, int prev_publish_time, int ema_price, int ema_conf, slice new_cs) = read_and_verify_message(cs, root_digest);
         cs = new_cs;
-        int should_include = (price_ids.tlen() == 0)
-            | ((price_ids.tlen() > 0)
-            & (publish_time >= min_publish_time)
-            & (publish_time <= max_publish_time)
-            & ((unique == 0) | (min_publish_time > prev_publish_time)));
-
-        if (should_include) {
-            ;; Create price feed cell containing both current and EMA prices
-            cell price_feed_cell = begin_cell()
-                .store_ref(store_price(price, conf, expo, publish_time))
-                .store_ref(store_price(ema_price, ema_conf, expo, publish_time))
-            .end_cell();
 
-            tuple price_feed = empty_tuple();
-            price_feed~tpush(price_id);
-            price_feed~tpush(price_feed_cell);
-            price_feeds~tpush(price_feed);
-
-            ;; Mark this price ID as found
-            found_price_ids~udict_set(256, price_id, begin_cell().store_int(true, 1).end_cell().begin_parse());
+        int price_ids_len = price_ids.tlen();
+
+        ;; Check if we've already processed this price_id to avoid duplicates
+        (_, int already_processed?) = found_price_ids.udict_get?(256, price_id);
+        if (~ already_processed?) { ;; Only process if we haven't seen this price_id yet
+            int should_include = (price_ids_len == 0)
+                | ((price_ids_len > 0)
+                & (publish_time >= min_publish_time)
+                & (publish_time <= max_publish_time)
+                & ((unique == 0) | (min_publish_time > prev_publish_time)));
+
+            if (should_include) {
+                ;; Create price feed cell containing both current and EMA prices
+                cell price_feed_cell = begin_cell()
+                    .store_ref(store_price(price, conf, expo, publish_time))
+                    .store_ref(store_price(ema_price, ema_conf, expo, publish_time))
+                .end_cell();
+
+                if (price_ids_len == 0) {
+                    ordered_feeds~udict_set(8, index, begin_cell()
+                        .store_uint(price_id, 256)
+                        .store_ref(price_feed_cell)
+                    .end_cell().begin_parse());
+                    index += 1;
+                } else {
```

Comment:
This is probably when having a top level if to return in this case makes the code cleaner.